### ISSUE TITLE
poor qrnn performance

### ISSUE THREAD

**Issue by PiotrCzapla** [fastai/fastai#1108]
**Describe the bug**
as Sebastian Ruder described qrrn seems to be crippled and does not get the  as fast as LSTM

**To Reproduce**
I'm running test prepared by sebastian from our [ULMFit-multilingual](https://github.com/n-waves/ulmfit-multilingual) repository, LSTM and QRNN Should coverage to similar values after the same number of epochs, they do not.
Fetchwikitext-2 to data/en/wikitext-2
Run this command to train a QRNN on a limited data set (10% of wiki-2) 
```
time python -m ulmfit.pretrain_lm data/en/wikitext-2 --qrnn=True --cuda-id=1 --name=wt-2-qrrn --num_epochs=3 --ds-pct 0.1
Batch size: 70
Max vocab: 60000
Using QRNNs...
Size of vocabulary: 15246
First 10 words in vocab: the, <pad>, ,, ., of, <unk>, and, in, to, <eos>
Saving vocabulary...
epoch  train_loss  valid_loss  accuracy
1      8.567326    6.519516    0.056718
2      7.420971    6.487216    0.054253
3      7.061250    6.497565    0.054233
python -m ulmfit.pretrain_lm data/en/wikitext-2 --qrnn=True --cuda-id=1    0.  11.77s user 4.25s system 99% cpu 16.096 total
```

Run this command to train a LSTM on a limited data set (10% of wiki-2) 
```
time python -m ulmfit.pretrain_lm data/en/wikitext-2 --qrnn=False --name=wt-2 --num_epochs=3 --ds-pct=0.1
Batch size: 70
Max vocab: 60000
Limiting data set to 0.1
Size of vocabulary: 15246
First 10 words in vocab: the, <pad>, ,, ., of, <unk>, and, in, to, <eos>
Saving vocabulary...
epoch  train_loss  valid_loss  accuracy
1      7.324363    6.342090    0.122999
2      6.559756    5.782920    0.182868
3      6.085223    5.686380    0.195393
python -m ulmfit.pretrain_lm data/en/wikitext-2 --qrnn=False --name=wt-2    17.20s user 6.35s system 98% cpu 23.881 total
````

The loss should be similar but it isn't LSTM Gets to 5.68 val_loss while QRRN gets only to  6.49 (place where LSTM is after 1st epoch)

**Expected behavior**
LSTM and QRNN Should have similar performance but QRNN Should be faster - This is assumption i need to test that on the old version of fastai

I'm going to investigate the issue and submit a patch once I figure out what is going on.

**Comment by PiotrCzapla** [comment]
QRNN after 80 epochs
```
python -m ulmfit.pretrain_lm data/en/wikitext-2 --qrnn=True --cuda-id=1 --name=wt-2-qrrn --num_epochs=100 --ds-pct 0.1
Batch size: 70
Max vocab: 60000
Using QRNNs...
Limiting data set to 0.1
Size of vocabulary: 15246
First 10 words in vocab: the, <pad>, ,, ., of, <unk>, and, in, to, <eos>
Saving vocabulary...
epoch  train_loss  valid_loss  accuracy
1      9.589046    9.525568    0.054306
2      9.424072    9.135739    0.054414
3      9.018161    8.359212    0.055189
4      8.346138    7.305996    0.053968
5      7.580266    6.492858    0.055856
6      7.169321    6.493022    0.054306
7      6.987562    6.489878    0.054160
8      6.907942    6.492623    0.053957
9      6.873025    6.489362    0.055335
10     6.859171    6.484140    0.054791
11     6.853229    6.491655    0.054306
12     6.848705    6.489232    0.054306
13     6.845232    6.473944    0.055195
14     6.842732    6.485246    0.062665
15     6.839021    6.484037    0.054306
16     6.830321    6.465160    0.071581
17     6.818776    6.452264    0.075586
18     6.800464    6.431604    0.087741
19     6.767688    6.376600    0.095410
20     6.724560    6.318267    0.103209
21     6.671863    6.255381    0.115368
22     6.606373    6.166226    0.132358
23     6.535790    6.103925    0.138655
24     6.459231    6.043802    0.149233
25     6.385064    6.010012    0.152178
26     6.301238    5.932988    0.160843
27     6.211205    5.891285    0.165704
28     6.124407    5.835395    0.176596
29     6.032049    5.793059    0.174895
30     5.943070    5.763924    0.180171
31     5.846730    5.705569    0.190848
32     5.744658    5.666500    0.192437
33     5.650591    5.643042    0.194758
34     5.554672    5.609411    0.199240
35     5.447789    5.580114    0.201429
36     5.351630    5.552900    0.202304
37     5.273030    5.564426    0.203850
38     5.174746    5.534225    0.205236
39     5.078378    5.522253    0.207350
40     5.000897    5.530851    0.206839
41     4.921390    5.521262    0.207447
42     4.841522    5.530956    0.207143
43     4.763840    5.558990    0.206618
44     4.686889    5.555052    0.206436
45     4.606619    5.539403    0.207801
46     4.535450    5.544782    0.206550
47     4.471734    5.551255    0.208156
48     4.413557    5.582464    0.205978
49     4.357163    5.581117    0.205876
50     4.293034    5.580110    0.207778
51     4.246437    5.609156    0.204474
52     4.227892    5.614207    0.203901
53     4.167626    5.608902    0.205067
54     4.098423    5.591347    0.206855
55     4.051802    5.634832    0.201623
56     4.025866    5.644625    0.202950
57     3.960732    5.628582    0.203588
58     3.919922    5.643956    0.204407
59     3.871385    5.649668    0.203597
60     3.863034    5.665218    0.204186
61     3.822766    5.656523    0.203662
62     3.784243    5.667356    0.205397
63     3.758609    5.682689    0.202189
64     3.752130    5.687783    0.202890
65     3.725233    5.684005    0.204711
66     3.680005    5.693838    0.202549
67     3.644393    5.680868    0.202727
68     3.612632    5.688258    0.201240
69     3.618646    5.703245    0.202049
70     3.579999    5.689683    0.200999
71     3.552939    5.712533    0.199179
72     3.531558    5.715834    0.200253
73     3.515812    5.735734    0.196589
74     3.515083    5.728702    0.198401
75     3.520288    5.731945    0.199544
76     3.485003    5.731506    0.200362
77     3.478617    5.735614    0.198884
78     3.469165    5.737576    0.200473
79     3.476101    5.741198    0.199635
80     3.458618    5.726952    0.200059
```
LSTM after 80 epochs
```
time python -m ulmfit.pretrain_lm data/en/wikitext-2 --qrnn=False --name=wt-2 --num_epochs=100 --ds-pct=0.1
Batch size: 70
Max vocab: 60000
Limiting data set to 0.1
Size of vocabulary: 15246
First 10 words in vocab: the, <pad>, ,, ., of, <unk>, and, in, to, <eos>
Saving vocabulary...
epoch  train_loss  valid_loss  accuracy
1      8.149413    6.625456    0.062006
2      7.230876    6.325223    0.099559
3      6.830673    6.125136    0.144649
4      6.551603    5.952240    0.173658
5      6.325869    5.836593    0.185208
6      6.125107    5.735621    0.195029
7      5.946550    5.717289    0.202178
8      5.771859    5.629531    0.199899
9      5.599160    5.556547    0.207811
10     5.419254    5.529798    0.209928
11     5.225041    5.507531    0.213010
12     5.034083    5.549269    0.205559
13     4.824013    5.621671    0.199627
14     4.613435    5.691539    0.196902
15     4.409591    5.702592    0.193010
16     4.195541    5.784494    0.195985
17     4.038594    5.646560    0.203244
18     3.841888    5.732875    0.195725
19     3.608887    6.036916    0.180364
20     3.372380    6.232230    0.168756
21     3.128355    6.262933    0.176038
22     2.909338    6.441030    0.171536
23     2.674638    6.417610    0.171429
24     2.433226    6.704102    0.169098
25     2.277819    6.463394    0.169116
26     1.988678    7.060129    0.154458
27     1.784888    7.097499    0.155542
28     1.557777    7.144700    0.156282
29     1.406564    7.181555    0.152254
30     1.199157    7.481484    0.152097
31     1.035858    7.643167    0.130978
32     0.892027    7.752474    0.135309
33     0.734421    7.835570    0.148244
34     0.650520    7.789397    0.144553
35     0.535672    7.819457    0.145871
36     0.479333    7.990303    0.147565
37     0.445199    8.043110    0.142208
38     0.372400    8.086502    0.136321
39     0.362246    8.095461    0.138166
40     0.297715    8.294946    0.137842
41     0.269403    8.398767    0.140062
42     0.231488    8.379242    0.137079
43     0.328645    8.302129    0.138708
44     0.301683    8.359775    0.135945
45     0.228592    8.246195    0.139716
46     0.187372    8.308817    0.139564
47     0.235492    8.342047    0.139108
48     0.196987    8.322741    0.140780
49     0.164636    8.369329    0.144128
50     0.132969    8.272486    0.142918
51     0.130011    8.361520    0.137639
52     0.109054    8.150353    0.147692
53     0.112340    8.291537    0.139260
54     0.277651    7.856262    0.147011
55     0.171290    8.075988    0.146454
56     0.114004    8.053049    0.146480
57     0.092124    8.044921    0.146701
58     0.088315    8.128678    0.145542
59     0.075355    8.009786    0.148571
60     0.081272    7.991426    0.145947
61     0.085800    8.081589    0.142396
62     0.069526    7.963127    0.141535
63     0.062828    7.908299    0.148162
64     0.069758    7.912842    0.147163
65     0.070617    7.826627    0.151533
66     0.057702    7.805183    0.149255
67     0.052341    7.751988    0.150462
68     0.047684    7.770833    0.150185
69     0.047177    7.793154    0.149032
70     0.047017    7.764704    0.151601
71     0.062799    7.766047    0.153231
72     0.057702    7.762446    0.152229
73     0.047140    7.732766    0.152135
74     0.042476    7.683928    0.154945
75     0.048810    7.631882    0.154481
76     0.107342    7.568276    0.155231
77     0.073181    7.645492    0.154205
78     0.049275    7.666656    0.153749
79     0.060635    7.702621    0.152312
80     0.057178    7.706323    0.151134
```

**Comment by PiotrCzapla** [comment]
Funny enough the same code that does not work from console works fine in jupyter notebook :), and QRNN trains properly.
<img width="940" alt="screenshot 2018-11-09 at 00 35 36" src="https://user-images.githubusercontent.com/340180/48233902-6c485480-e3b7-11e8-857e-cfa7472f56ab.png">

**Comment by sgugger** [comment]
It should be fixed now.

**Comment by PiotrCzapla** [comment]
Unfortunately, it isn't fixed yet. If you run `fit` twice the issue disappears for the second run. I think the issue was there from the start, but in old fastai fit was working a bit differently. 
Have a look at the screenshot below. It is on current fastai (updated 10 min ago)
<img width="689" alt="screenshot 2018-11-09 at 22 38 27" src="https://user-images.githubusercontent.com/340180/48289596-50ed5000-e470-11e8-9cc8-fa46001fbb02.png">

**Comment by PiotrCzapla** [comment]
To be exact, second run to create_opt fixes the issue 
<img width="947" alt="screenshot 2018-11-09 at 22 53 18" src="https://user-images.githubusercontent.com/340180/48290200-53e94000-e472-11e8-9940-0e8b12874296.png">

**Comment by sgugger** [comment]
Ah, you're definitely on to something. I think there needs to be a `learn.model.reset()` *before* the optimizer is created.
Will look into this.

**Comment by PiotrCzapla** [comment]
@sgugger  I've added some tests to show how this issue can be reproduced
